### PR TITLE
Add EmitC support for comparison ops

### PR DESCRIPTION
**Pull request by dajuro**
Further adds tests for `vm.cmp` ops to the VM. Note: pseudo-ops are not yet supported by EmitC. So no tests for these operations.

### CODE REVIEW

File: iree/vm/ops.h [hunk 1]
```diff
@@ -57,6 +57,24 @@ static inline int32_t vm_shr_i32u(int32_t operand, int8_t amount) {
...
+  return a == b ? 1 : 0;
+}
+static inline bool vm_cmp_ne_i32(int32_t a, int32_t b) {
+  return a != b ? 1 : 0;
+}
+static inline bool vm_cmp_lt_i32s(int32_t a, int32_t b) {
+  return a < b ? 1 : 0;
+}
+static inline bool vm_cmp_lt_i32u(uint32_t a, uint32_t b) {
+  return a < b ? 1 : 0;
```
Comment (marbre) [inline iree/vm/ops.h:74]:
```suggestion
static inline bool vm_cmp_lt_i32u(int32_t lhs, int32_t rhs) {
  (((uint32_t)lhs)<((uint32_t)rhs)) ? 1 : 0;

Comment (marbre) [reply iree/vm/ops.h:74]: Please make sure that our functions always consume and return `int32_t`.

File: iree/vm/ops.h [hunk 2]
```diff
@@ -57,6 +57,24 @@ static inline int32_t vm_shr_i32u(int32_t operand, int8_t amount) {
   return (int32_t)(((uint32_t)operand) >> amount);
 };
 
+//===------------------------------------------------------------------===//
+// Comparison ops
+//===------------------------------------------------------------------===//
+
+static inline bool vm_cmp_eq_i32(int32_t a, int32_t b) {
+  return a == b ? 1 : 0;
```
Comment (marbre) [inline iree/vm/ops.h:65]:
```suggestion
static inline bool vm_cmp_eq_i32(int32_t lhs, int32_t rhs) {
  return lhs == rhs ? 1 : 0;
```

I would prefer to make the functions look more consistent, meaning to use `lhs` and `rhs`. In addition, we could be closer to the original `(((type)lhs)op((type)rhs)) ? 1 : 0;` pattern, meaning to only skip superfluous `(type)` casts. This applies to the functions below as well.

File: iree/vm/ops.h [hunk 3]
```diff
@@ -57,6 +57,24 @@ static inline int32_t vm_shr_i32u(int32_t operand, int8_t amount) {
...
+static inline bool vm_cmp_ne_i32(int32_t a, int32_t b) {
+  return a != b ? 1 : 0;
+}
+static inline bool vm_cmp_lt_i32s(int32_t a, int32_t b) {
+  return a < b ? 1 : 0;
+}
+static inline bool vm_cmp_lt_i32u(uint32_t a, uint32_t b) {
+  return a < b ? 1 : 0;
+}
+static inline bool vm_cmp_nz_i32(int32_t a) { return (a != 0) ? 1 : 0; }
```
Comment (marbre) [inline iree/vm/ops.h:76]:
```suggestion
static inline bool vm_cmp_nz_i32(int32_t operand) { return (operand != 0) ? 1 : 0; }
```

File: iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/comparison_ops.mlir [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+// RUN: iree-opt -split-input-file -pass-pipeline='vm.module(iree-convert-vm-to-emitc)' %s | IreeFileCheck %s
+
+// CHECK: vm.module @module {
+vm.module @module {
+  // CHECK-LABEL: vm.func @cmp_eq_i32
+  vm.func @cmp_eq_i32(%arg0 : i32, %arg1 : i32) -> i32 {
+    // CHECK-NEXT: %0 = emitc.call "vm_cmp_eq_i32"(%arg0, %arg1) {args = [0 : index, 1 : index]} : (i32, i32) -> i32
+    %0 = vm.cmp.eq.i32 %arg0, %arg1 : i32
+    // CHECK-NEXT: vm.return
```
Comment (marbre) [inline iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/comparison_ops.mlir:11]:
```suggestion
```
Same as above. There is no real benefit as we only want to test the conversion.

File: iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/comparison_ops.mlir [hunk 5]
```diff
@@ -0,0 +1,68 @@
+// Tests printing and parsing of comparison ops.
+
+// RUN: iree-opt -split-input-file -pass-pipeline='vm.module(iree-convert-vm-to-emitc)' %s | IreeFileCheck %s
+
+// CHECK: vm.module @module {
```
Comment (marbre) [inline iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/comparison_ops.mlir:5]:
```suggestion
```

No need to check this.

**Review by marbre — CHANGES_REQUESTED** [review]

File: iree/vm/ops.h (hunk 1 above)
Comment (dajuro) [reply iree/vm/ops.h]: Fixed with latest push

File: iree/vm/ops.h (hunk 2 above)
Comment (dajuro) [reply iree/vm/ops.h]: Fixed with latest push

File: iree/vm/ops.h (hunk 3 above)
Comment (dajuro) [reply iree/vm/ops.h]: Fixed with latest push

File: iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/comparison_ops.mlir (hunk 4 above)
Comment (dajuro) [reply iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/comparison_ops.mlir]: Fixed with latest push

File: iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/comparison_ops.mlir (hunk 5 above)
Comment (dajuro) [reply iree/compiler/Dialect/VM/Conversion/VMToEmitC/test/comparison_ops.mlir]: Fixed with latest push

**Review by marbre — APPROVED** [review]
LGTM!

**Review by benvanik — APPROVED** [review]

**Comment by GMNGeoffrey** [comment]
So something really weird happened with the CLA here. All the commits on this branch are from David's fraunhofer address, but the squash commit is from a gmail address, which hasn't signed the CLA. Normally we wouldn't even notice, but when we merged into the google branch the CLA check complained (https://github.com/google/iree/pull/4811). This isn't actually a problem from a policy perspective I think. It's just annoying noise on the checks

@benvanik did you get any sort of dropdown for email to select when you squashed? I'm guessing not and this is just yet another instance of https://github.com/todogroup/gh-issues/issues/32

@dajuro can you confirm that your gmail is your default email on GitHub? I think that's what it uses.

**Pull request merged**